drivers/overlay/overlay.go

```diff
@@ -1175,19 +1179,27 @@ func (d *Driver) getLower(parent string) (string, error) {
 }
 
 func (d *Driver) dir(id string) string {
-	p, _, _ := d.dir2(id, false)
+	p, _, _, _ := d.dir2(id, false, false)
```

Comment:
So it seems that https://github.com/containers/storage/pull/1899/files#r1576924079 was resolved by just not _finding_ a VFS layer when `dir` is called, and instead returning an otherwise-uninitialized `primaryPath`. So some callers looking for a layer with the same ID will see a VFS layer, and some will see a different, empty directory?

Why is that safe and consistent?

After looking a bit, it seems that only parent layers can be VFS layers (because non-parent layers would presumably be handled by the VFS driver directly, not overlay?) … but then `isParent` is calling `.dir` and not `dir2`, and it will not find VFS layers?

---

I’m sorry, I just can’t deal with the complexity. Too many options, too many interactions, too many things that _seem_ like they might be possible even if something elsewhere in the call stack probably prevents them. **Most of that with documentation of neither intent nor maintained invariants**.

I straight can’t approve any PR adding a feature adding any more indirections or alternatives to the overlay driver any more. I’m not smart enough to keep up.

It seems to me that most non-trivial PRs here would require me to spend something like 2 days just graphing out and understanding the interactions and possibilities. At some point, it must pay off to instead spend 2-3 weeks to restructure and explicitly document, so that PRs are again a matter of a small number of hours, and I’m now past the point where I’m willing to enable the current state to persist, and to keep paying for the complexity those 2 days at a time, every time.

I don’t know what is the path forward. Probably easiest, find some other reviewer who is up to the task.

Or, maybe, aggressively simplify. Build abstractions where they can be built and don’t leak. Find identifiable an isolated utilities that can be cut out of the >400-line functions. Rename things to be clear (`dir` vs. `dir2` could never be clear), maybe `findParentLayerPossiblyInOtherDriver` vs. `findLayerWeOwn` vs … (yes, that is surely unproductive overkill, but the naming needs to _first_ be clear and only _afterwards_ succinct [I have no idea whether that’s the distinction that matters here, those names are not intended to be just used because I had to come up with an example]).

---

*shrug* it’s also possible that I’m overreacting and that a few sentences can make the picture clear. I’m not smart enough to figure that out myself.